### ISSUE TITLE
Template variables are executing excessive queries on InfluxDB

### ISSUE THREAD

**Issue by joshuajoh** [grafana/grafana#11890]
### What Grafana version are you using?

v5.0.3

### What datasource are you using?

InfluxDB 1.5.1

### What OS are you running grafana on?

RHEL 6.8

### What did you do?

Created a dashboard with 6 template variables.  

### What was the expected result?

Grafana should execute 6 "SHOW TAG VALUES" to populate the variables.

### What happened instead?

Grafana fired 32 "SHOW TAG VALUES" on the InfluxDB to populate the variables.  Many of the extra queries sent did not have values populated for the variables.  InfluxDB log snippet is below.

Please let me know if any other information is required.  Thank-you.

![grafana_influxdb_show](https://user-images.githubusercontent.com/22777878/39786497-7b3cfdb8-52e6-11e8-89c9-b60c423851f4.JPG)

[grafana_dashboard.zip](https://github.com/grafana/grafana/files/1985844/grafana_dashboard.zip)

### Sample Data for InfluxDB:
 
```
> create database temp_show
> grant read on temp_show to grafana
> use temp_show
Using database temp_show
> INSERT temp,t1=1,t2=2,t3=3,t4=4,t5=5,t6=6,t7=7,t8=8,t9=9,t10=10 txt="sample lp"
> INSERT temp,t1=a,t2=b,t3=c,t4=d,t5=e,t6=f,t7=g,t8=h,t9=i,t10=j txt="sample lp"
> select * from temp
name: temp
time                t1 t10 t2 t3 t4 t5 t6 t7 t8 t9 txt
----                -- --- -- -- -- -- -- -- -- -- ---
1525817387367675238 1  10  2  3  4  5  6  7  8  9  sample lp
1525817391444704872 a  j   b  c  d  e  f  g  h  i  sample lp
>
```

### InfluxDB log

```
ts=2018-05-08T22:17:44.077717Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t2 WHERE (_name = 'temp') AND ((t1 = 'a') AND (_tagKey = 't2'))"
ts=2018-05-08T22:17:44.094226Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t3 WHERE (_name = 'temp') AND ((t2 = '2' AND t1 = 'a') AND (_tagKey = 't3'))"
ts=2018-05-08T22:17:44.094924Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t4 WHERE (_name = 'temp') AND ((t3 = '3' AND t2 = '2' AND t1 = 'a') AND (_tagKey = 't4'))"
ts=2018-05-08T22:17:44.095967Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t5 WHERE (_name = 'temp') AND ((t4 = '4' AND t3 = '3' AND t2 = '2' AND t1 = 'a') AND (_tagKey = 't5'))"
ts=2018-05-08T22:17:44.099482Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t6 WHERE (_name = 'temp') AND ((t5 = '5' AND t4 = '4' AND t3 = '3' AND t2 = '2' AND t1 = 'a') AND (_tagKey = 't6'))"
ts=2018-05-08T22:17:44.414153Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t3 WHERE (_name = 'temp') AND ((t2 = 'b' AND t1 = 'a') AND (_tagKey = 't3'))"
ts=2018-05-08T22:17:44.414333Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t4 WHERE (_name = 'temp') AND ((t3 = '' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't4'))"
ts=2018-05-08T22:17:44.415193Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t5 WHERE (_name = 'temp') AND ((t4 = '' AND t3 = '' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't5'))"
ts=2018-05-08T22:17:44.416265Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t6 WHERE (_name = 'temp') AND ((t5 = '' AND t4 = '' AND t3 = '' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't6'))"
ts=2018-05-08T22:17:44.416615Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t5 WHERE (_name = 'temp') AND ((t4 = '' AND t3 = '' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't5'))"
ts=2018-05-08T22:17:44.416866Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t4 WHERE (_name = 'temp') AND ((t3 = '' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't4'))"
ts=2018-05-08T22:17:44.417799Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t6 WHERE (_name = 'temp') AND ((t5 = '' AND t4 = '' AND t3 = '' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't6'))"
ts=2018-05-08T22:17:44.418511Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t5 WHERE (_name = 'temp') AND ((t4 = '' AND t3 = '' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't5'))"
ts=2018-05-08T22:17:44.419033Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t6 WHERE (_name = 'temp') AND ((t5 = '' AND t4 = '' AND t3 = '' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't6'))"
ts=2018-05-08T22:17:44.420371Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t6 WHERE (_name = 'temp') AND ((t5 = '' AND t4 = '' AND t3 = '' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't6'))"
ts=2018-05-08T22:17:44.551993Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t5 WHERE (_name = 'temp') AND ((t4 = '' AND t3 = '' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't5'))"
ts=2018-05-08T22:17:44.562038Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t6 WHERE (_name = 'temp') AND ((t5 = '' AND t4 = '' AND t3 = '' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't6'))"
ts=2018-05-08T22:17:44.776560Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t4 WHERE (_name = 'temp') AND ((t3 = 'c' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't4'))"
ts=2018-05-08T22:17:44.777116Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t5 WHERE (_name = 'temp') AND ((t4 = '' AND t3 = 'c' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't5'))"
ts=2018-05-08T22:17:44.777542Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t6 WHERE (_name = 'temp') AND ((t5 = '' AND t4 = '' AND t3 = 'c' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't6'))"
ts=2018-05-08T22:17:44.777621Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t6 WHERE (_name = 'temp') AND ((t5 = '' AND t4 = '' AND t3 = 'c' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't6'))"
ts=2018-05-08T22:17:44.777877Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t6 WHERE (_name = 'temp') AND ((t5 = '' AND t4 = '' AND t3 = 'c' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't6'))"
ts=2018-05-08T22:17:44.809358Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t5 WHERE (_name = 'temp') AND ((t4 = '' AND t3 = 'c' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't5'))"
ts=2018-05-08T22:17:44.811355Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t6 WHERE (_name = 'temp') AND ((t5 = '' AND t4 = '' AND t3 = 'c' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't6'))"
ts=2018-05-08T22:17:44.815807Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t6 WHERE (_name = 'temp') AND ((t5 = '' AND t4 = '' AND t3 = 'c' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't6'))"
ts=2018-05-08T22:17:44.891917Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t6 WHERE (_name = 'temp') AND ((t5 = '' AND t4 = '' AND t3 = 'c' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't6'))"
ts=2018-05-08T22:17:45.137826Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t5 WHERE (_name = 'temp') AND ((t4 = 'd' AND t3 = 'c' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't5'))"
ts=2018-05-08T22:17:45.138017Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t6 WHERE (_name = 'temp') AND ((t5 = '' AND t4 = 'd' AND t3 = 'c' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't6'))"
ts=2018-05-08T22:17:45.151029Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t6 WHERE (_name = 'temp') AND ((t5 = '' AND t4 = 'd' AND t3 = 'c' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't6'))"
ts=2018-05-08T22:17:45.189286Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t6 WHERE (_name = 'temp') AND ((t5 = '' AND t4 = 'd' AND t3 = 'c' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't6'))"
ts=2018-05-08T22:17:45.401452Z lvl=info msg="Executing query" log_id=07xCd73W000 service=query query="SHOW TAG VALUES ON temp_show WITH KEY = t6 WHERE (_name = 'temp') AND ((t5 = 'e' AND t4 = 'd' AND t3 = 'c' AND t2 = 'b' AND t1 = 'a') AND (_tagKey = 't6'))"
```

**Comment by torkelo** [comment]
strange, unable to replicate here:
https://play.grafana.org/d/000000002/influxdb-templated-queries?orgId=1&var-datacenter=Africa&var-host=All&var-summarize=1m 

What is you variable options and queres?

**Comment by joshuajoh** [comment]
Screenshots of variables and query are below.

Upon changing the first level variable, here are the number of executed shows:

6 variables:  31
5 variables:  15
4 variables:    7
3 variables:    3
2 variables:    1

Thank-you.

![variable_list](https://user-images.githubusercontent.com/22777878/39833529-e764ad84-538f-11e8-8646-b7072831d80a.JPG)
![variable_options](https://user-images.githubusercontent.com/22777878/39833533-e97a5452-538f-11e8-9c4d-547fde68d594.JPG)

**Comment by yellowpattern** [comment]
Does grafana do another set of 31 queries when you change any of the values?

**Comment by joshuajoh** [comment]
@yellowpattern 

Yes, 31 queries are fired whenever the Tier1 variable is changed on the 2 rows of sample data.

I'll populate some more sample data and test the number of SHOW executions for the Tier2-6 variables later today.

**Comment by marefr** [comment]
@joshuajoh do you have Refresh on dashboard load for all of the variables?

**Comment by joshuajoh** [comment]
@marefr yes, that option is set for each variable.

**Comment by marefr** [comment]
Have not been able to reproduce this in latest master.

Also tried v5.0.3 using this [dashboard](https://k8s.devstats.cncf.io/d/5/bot-commands-repository-groups?orgId=1) but when the dashboard loads only 4 SHOW TAG VALUES queries are executed.

Will close this, but consider reopening this if you can provide us with a working example or guideline for setting up a dashboard having the behavior you're referring to.

**Comment by joshuajoh** [comment]
@marefr i looked at the [dashboard](https://k8s.devstats.cncf.io/d/5/bot-commands-repository-groups?orgId=1) and it since it only has 2 nested/dependent variables there will be no extra SHOW commands executed.

The behavior of Grafana sending extra SHOW queries with empty values starts when there more than 2 nested/dependent variables.

The sample data and the dashboard file in the first post can be used to replicate this.

Is it possible that Grafana only support 2 nested variables?

Thank-you.

**Comment by marefr** [comment]
@joshuajoh Finally got the the time to investigate this using for excellent instruction. It's clearly a bug and something strange is going on.

**Variables:**
Tier1=a
Tier2=2
Tier3=3
Tier4=4
Tier5=5
Tier6=6
Changing Tier1=1

**Requests:**
```
SHOW TAG VALUES FROM temp with key = "t2" WHERE t1 = '1'
SHOW TAG VALUES FROM temp with key = "t3" WHERE t1 = '1' AND t2 = 'b'
SHOW TAG VALUES FROM temp with key = "t4" WHERE t1 = '1' AND t2 = 'b' AND t3 = 'c'
SHOW TAG VALUES FROM temp with key = "t5" WHERE t1 = '1' AND t2 = 'b' AND t3 = 'c' AND t4 = 'd'
SHOW TAG VALUES FROM temp with key = "t6" WHERE t1 = '1' AND t2 = 'b' AND t3 = 'c' AND t4 = 'd' AND t5 = 'e'
SHOW TAG VALUES FROM temp with key = "t3" WHERE t1 = '1' AND t2 = '2'
SHOW TAG VALUES FROM temp with key = "t4" WHERE t1 = '1' AND t2 = '2' AND t3 = ''
SHOW TAG VALUES FROM temp with key = "t5" WHERE t1 = '1' AND t2 = '2' AND t3 = '' AND t4 = 'd'
SHOW TAG VALUES FROM temp with key = "t6" WHERE t1 = '1' AND t2 = '2' AND t3 = '' AND t4 = 'd' AND t5 = 'e'
SHOW TAG VALUES FROM temp with key = "t4" WHERE t1 = '1' AND t2 = '2' AND t3 = ''
SHOW TAG VALUES FROM temp with key = "t5" WHERE t1 = '1' AND t2 = '2' AND t3 = '' AND t4 = 'd'
SHOW TAG VALUES FROM temp with key = "t6" WHERE t1 = '1' AND t2 = '2' AND t3 = '' AND t4 = 'd' AND t5 = 'e'
SHOW TAG VALUES FROM temp with key = "t6" WHERE t1 = '1' AND t2 = '2' AND t3 = '3' AND t4 = '' AND t5 = ''
SHOW TAG VALUES FROM temp with key = "t5" WHERE t1 = '1' AND t2 = '2' AND t3 = '3' AND t4 = ''
SHOW TAG VALUES FROM temp with key = "t6" WHERE t1 = '1' AND t2 = '2' AND t3 = '3' AND t4 = '' AND t5 = ''
SHOW TAG VALUES FROM temp with key = "t4" WHERE t1 = '1' AND t2 = '2' AND t3 = '3'
SHOW TAG VALUES FROM temp with key = "t5" WHERE t1 = '1' AND t2 = '2' AND t3 = '3' AND t4 = ''
SHOW TAG VALUES FROM temp with key = "t6" WHERE t1 = '1' AND t2 = '2' AND t3 = '3' AND t4 = '' AND t5 = ''
SHOW TAG VALUES FROM temp with key = "t5" WHERE t1 = '1' AND t2 = '2' AND t3 = '3' AND t4 = ''
SHOW TAG VALUES FROM temp with key = "t6" WHERE t1 = '1' AND t2 = '2' AND t3 = '3' AND t4 = '' AND t5 = ''
SHOW TAG VALUES FROM temp with key = "t6" WHERE t1 = '1' AND t2 = '2' AND t3 = '3' AND t4 = '' AND t5 = ''
SHOW TAG VALUES FROM temp with key = "t5" WHERE t1 = '1' AND t2 = '2' AND t3 = '3' AND t4 = '4'
SHOW TAG VALUES FROM temp with key = "t6" WHERE t1 = '1' AND t2 = '2' AND t3 = '3' AND t4 = '4' AND t5 = ''
SHOW TAG VALUES FROM temp with key = "t6" WHERE t1 = '1' AND t2 = '2' AND t3 = '3' AND t4 = '4' AND t5 = ''
SHOW TAG VALUES FROM temp with key = "t6" WHERE t1 = '1' AND t2 = '2' AND t3 = '3' AND t4 = '4' AND t5 = ''
SHOW TAG VALUES FROM temp with key = "t6" WHERE t1 = '1' AND t2 = '2' AND t3 = '3' AND t4 = '4' AND t5 = ''
SHOW TAG VALUES FROM temp with key = "t5" WHERE t1 = '1' AND t2 = '2' AND t3 = '3' AND t4 = '4'
SHOW TAG VALUES FROM temp with key = "t6" WHERE t1 = '1' AND t2 = '2' AND t3 = '3' AND t4 = '4' AND t5 = ''
SHOW TAG VALUES FROM temp with key = "t6" WHERE t1 = '1' AND t2 = '2' AND t3 = '3' AND t4 = '4' AND t5 = ''
SHOW TAG VALUES FROM temp with key = "t6" WHERE t1 = '1' AND t2 = '2' AND t3 = '3' AND t4 = '4' AND t5 = '5'
SHOW TAG VALUES FROM temp with key = "t6" WHERE t1 = '1' AND t2 = '2' AND t3 = '3' AND t4 = '4' AND t5 = '5'
```

**Comment by parisnps** [comment]
I am seeing similar behavior with template variable using Elasticsearch as backend.  I have similarly siz template variable which are each nested to one other. Is this bug & fix generic or only focussed on Influxdb as backend ?

**Comment by torkelo** [comment]
It’s s general problem, so fix will solve your case as well

**Comment by parisnps** [comment]
Has there been any progress on this ? Sorry to bother but this is a problem we end up noticing every day using grafana  ...

**Comment by marefr** [comment]
@parisnps not yet. It's currently targeted to be fixed and included in v5.3 version (no release date yet)
